File: Python/108. Convert Sorted Array to Binary Search Tree.py
'''
1.递归
2.迭代
'''
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

# class Solution:
#     def sortedArrayToBST(self, nums: List[int]) -> Optional[TreeNode]:
#         def createBST(nums, left, right):
#             if left >= right:
#                 return None
#             mid = (left + right) // 2
#             node = TreeNode(nums[mid])
#             node.left = createBST(nums, left, mid)
#             node.right = createBST(nums, mid+1, right)
#             return node
#         return createBST(nums, 0, len(nums))

class Solution:
    def sortedArrayToBST(self, nums: List[int]) -> Optional[TreeNode]:
        root = TreeNode(0)
        que = [root]
        leftQue = [0]
        rightQue = [len(nums) - 1]
        while que:
            node = que.pop(0)
            left = leftQue.pop(0)
            right = rightQue.pop(0)
            mid = (left + right) // 2
            node.val = nums[mid]
            if left < mid:
                node.left = TreeNode(0)
                que.append(node.left)
                leftQue.append(left)
                rightQue.append(mid-1)
            if mid < right:
                node.right = TreeNode(0)
                que.append(node.right)
                leftQue.append(mid+1)
                rightQue.append(right)
        return root

File: C++/108. Convert Sorted Array to Binary Search Tree.cpp
/*
1.µÝ¹é
2.µü´ú
*/
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

 // class Solution {
 // public:
 //     TreeNode* createBST(vector<int>& nums, int begin, int end){
 //         if (begin >= end) return NULL;
 //         int mid = (begin + end) / 2;
 //         TreeNode* node = new TreeNode(nums[mid]);
 //         node->left = createBST(nums, begin, mid);
 //         node->right = createBST(nums, mid+1, end);
 //         return node;
 //     }
 //     TreeNode* sortedArrayToBST(vector<int>& nums) {
 //         return createBST(nums, 0, nums.size());
 //     }
 // };

class Solution {
public:
    TreeNode* sortedArrayToBST(vector<int>& nums) {
        if (nums.size() == 0) return NULL;
        queue<TreeNode*> que;
        TreeNode* root = new TreeNode(0);
        que.push(root);
        queue<int> leftQue;
        queue<int> rightQue;
        leftQue.push(0);
        rightQue.push(nums.size() - 1);
        while (!que.empty()) {
            TreeNode* node = que.front(); que.pop();
            int left = leftQue.front(); leftQue.pop();
            int right = rightQue.front(); rightQue.pop();
            int mid = left + (right - left) / 2;
            node->val = nums[mid];
            if (left < mid) {
                node->left = new TreeNode(0);
                que.push(node->left);
                leftQue.push(left);
                rightQue.push(mid - 1);
            }
            if (mid < right) {
                node->right = new TreeNode(0);
                que.push(node->right);
                leftQue.push(mid + 1);
                rightQue.push(right);
            }
        }
        return root;
    }
};